x-pack/test_serverless/shared/services/user_manager/saml_auth.ts

```diff
@@ -0,0 +1,171 @@
+}
+
+const getSessionCookie = (cookieString: string) => {
+  return parseCookie(cookieString);
+};
+
+const getCloudHostName = () => {
+  const hostname = process.env.TEST_CLOUD_HOST_NAME;
+  if (!hostname) {
+    throw new Error('SAML Authentication requires TEST_CLOUD_HOST_NAME env variable to be set');
+  }
+
+  return hostname;
+};
+
+const getCloudUrl = (hostname: string, pathname: string) => {
+  return Url.format({
+    protocol: 'https',
+    hostname,
+    pathname,
+  });
+};
+
+const createCloudSession = async (hostname: string, email: string, password: string) => {
+  const cloudLoginUrl = getCloudUrl(hostname, '/api/v1/users/_login');
+  const sessionResponse: AxiosResponse = await axios.request({
+    url: cloudLoginUrl,
+    method: 'post',
+    data: {
+      email,
+      password,
+    },
+    headers: {
+      accept: 'application/json',
+      'content-type': 'application/json',
+    },
+    validateStatus: () => true,
+    maxRedirects: 0,
+  });
+  const firstName = sessionResponse?.data?.user?.data?.first_name ?? '';
+  const lastName = sessionResponse?.data?.user?.data?.last_name ?? '';
+  const fullname = `${firstName} ${lastName}`.trim();
+  return {
+    token: sessionResponse?.data?.token as string,
+    fullname,
+  };
+};
+
+const createSAMLRequest = async (kbnUrl: string, kbnVersion: string) => {
+  const samlResponse: AxiosResponse = await axios.request({
+    url: kbnUrl + '/internal/security/login',
+    method: 'post',
+    data: {
+      providerType: 'saml',
+      providerName: 'cloud-saml-kibana',
+      currentURL: kbnUrl + '/login?next=%2F"',
+    },
+    headers: {
+      'kbn-version': kbnVersion,
+      'x-elastic-internal-origin': 'Kibana',
+      'content-type': 'application/json',
+    },
+    validateStatus: () => true,
+    maxRedirects: 0,
+  });
+  const cookie = getSessionCookie(samlResponse.headers['set-cookie']![0])!;
+  return { location: samlResponse.data.location, sid: cookie.value };
+};
+
+const createSAMLResponse = async (url: string, ecSession: string) => {
+  const samlResponse = await axios.get(url, {
+    headers: {
+      Cookie: `ec_session=${ecSession}`,
+    },
+  });
+  const $ = cheerio.load(samlResponse.data);
+  const value = $('input').attr('value') ?? '';
+  if (value.length === 0) {
+    throw new Error('Failed to parse SAML response value');
+  }
+  return value;
+};
+
+const finishSAMLHandshake = async ({
+  kbnHost,
+  samlResponse,
+  sid,
+  log,
+}: {
+  kbnHost: string;
+  samlResponse: string;
+  sid?: string;
+  log: ToolingLog;
+}) => {
+  const encodedResponse = encodeURIComponent(samlResponse);
+  let authResponse: AxiosResponse;
+
+  try {
+    authResponse = await axios.request({
+      url: kbnHost + '/api/security/saml/callback',
+      method: 'post',
+      data: `SAMLResponse=${encodedResponse}`,
+      headers: {
+        ...(sid ? { Cookie: `sid=${sid}` } : {}),
+        ...{ 'content-type': 'application/x-www-form-urlencoded' },
+      },
```

Comment:
__nit:__
```suggestion
      headers: {
        'content-type': 'application/x-www-form-urlencoded',
        ...(sid ? { Cookie: `sid=${sid}` } : {}),
      },
```

Also wondering, maybe can be simplified a bit (haven't checked):
```ts
    authResponse = await axios.post(
      kbnHost + '/api/security/saml/callback',
      `SAMLResponse=${encodedResponse}`,
      {
        headers: {
          'content-type': 'application/x-www-form-urlencoded',
          ...(sid ? { Cookie: `sid=${sid}` } : {}),
        },
        validateStatus: () => true,
        maxRedirects: 0,
      }
    );
```